src/main/scala/org/broadinstitute/hail/asm4s/FunctionBuilder.scala

```diff
@@ -88,110 +88,402 @@ abstract class FunctionBuilder[R](parameterTypeInfo: Array[TypeInfo[_]], returnT
+
+  cn.interfaces.asInstanceOf[java.util.List[String]].add("org/broadinstitute/hail/asm4s/ZToI")
+
+  def arg1 = getArg[Boolean](1)
+
+  def result(c: Code[Int]): (Boolean) => Int = {
+    val bytes = classAsBytes(c)
+    val klass = act.runtimeClass
+    // FIXME: thread-safe?
+    val localName = name.replaceAll("/",".")
+
+    new Function1[Boolean, Int] with java.io.Serializable {
+      @transient @volatile private var f: ZToI = null
+      def apply(a: Boolean): Int = {
+        if (f == null) {
+          this.synchronized {
+            if (f == null) {
+              f = loadClass(localName, bytes).newInstance().asInstanceOf[ZToI]
+            }
+          }
+        }
+
+        f(a)
+      }
+    }
+  }
+}
 
-  def arg2 = getArg[A2](1)
+trait IToI { def apply(b: Int): Int }
+class FunctionIToIBuilder(implicit act: ClassTag[Int], ati: TypeInfo[Int],
+  rti: TypeInfo[Int]) extends FunctionBuilder[Int](Array[TypeInfo[_]](ati), rti) {
+
+  cn.interfaces.asInstanceOf[java.util.List[String]].add("org/broadinstitute/hail/asm4s/IToI")
+
+  def arg1 = getArg[Int](1)
+
+  def result(c: Code[Int]): (Int) => Int = {
+    val bytes = classAsBytes(c)
+    val localName = name.replaceAll("/",".")
+
+    new Function1[Int, Int] with java.io.Serializable {
+      @transient @volatile private var f: IToI = null
+      def apply(a: Int): Int = {
+        if (f == null) {
+          this.synchronized {
+            if (f == null) {
+              f = loadClass(localName, bytes).newInstance().asInstanceOf[IToI]
+            }
+          }
+        }
+
+        f(a)
+      }
+    }
+  }
+}
+
+
+trait AToI[A] { def apply(b: A): Int }
+class FunctionAToIBuilder[A](implicit act: ClassTag[A], ati: TypeInfo[A],
+  rti: TypeInfo[Int]) extends FunctionBuilder[Int](Array[TypeInfo[_]](implicitly[TypeInfo[AnyRef]]), rti) {
+
+  cn.interfaces.asInstanceOf[java.util.List[String]].add("org/broadinstitute/hail/asm4s/AToI")
+
+  def arg1 = Code.checkcast[A](getArg[AnyRef](1))
+
+  def result(c: Code[Int]): (A) => Int = {
+    val bytes = classAsBytes(c)
+    val localName = name.replaceAll("/",".")
+
+    new Function1[A, Int] with java.io.Serializable {
+      @transient @volatile private var f: AToI[A] = null
+      def apply(a: A): Int = {
+        if (f == null) {
+          this.synchronized {
+            if (f == null) {
+              f = loadClass(localName, bytes).newInstance().asInstanceOf[AToI[A]]
+            }
+          }
+        }
+
+        f(a)
+      }
+    }
+  }
+}
+
+class Function2Builder[A1 >: Null, A2 >: Null, R >: Null]
+  (implicit a1ct: ClassTag[A1], a1ti: TypeInfo[A1], a2ct: ClassTag[A2], a2ti: TypeInfo[A2], rti: TypeInfo[R])
+    extends FunctionBuilder[R](Array[TypeInfo[_]](implicitly[TypeInfo[AnyRef]], implicitly[TypeInfo[AnyRef]]), rti) {
+
+  cn.interfaces.asInstanceOf[java.util.List[String]].add("scala/Function2")
+
+  def arg1 = Code.checkcast[A1](getArg[AnyRef](1))
```

Comment:
The argument types are actually `AnyRef` in byte code because it implements a generic type, `Function2[T,U]`. The casts are necessary to get the correct type in byte code. Java deals with this by having two methods, one is the generic wrapper and the other is the primitive interface. This was easier for me to implement ;).